### PR TITLE
[WIP] Implements Roberta Model

### PR DESCRIPTION
**Pull request by JulienDarve**

### CODE REVIEW

File: src/levanter/models/testing.ipynb
Comment (dlwh) [inline src/levanter/models/testing.ipynb:1]: rm before merge

File: src/levanter/models/roberta.py [hunk 1]
```diff
@@ -0,0 +1,852 @@
...
+    cos: NamedArray,  # [position, head_size]
+    sin: NamedArray,  # [position, head_size]
+) -> Tuple[NamedArray, NamedArray]:
+    """Applies rotary position embedding to q and k."""
+    q_embed = q * cos + _rotate_half(q) * sin
+    k_embed = k * cos + _rotate_half(k) * sin
+    return q_embed, k_embed
+
+
+def llama_rotary_pos_emb(
```
Comment (dlwh) [inline src/levanter/models/roberta.py:836]: you'll want to delete this i think? (if you end up needing it, just import from llama.py)

File: src/levanter/models/roberta.py [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+
+        return RobertaForMaskedLM(roberta, lm_head, Vocab)
+
+    def get_output_embeddings(self):
+        return self.lm_head.decoder
+
+    def set_output_embeddings(self, new_embeddings):
+        self.lm_head.decoder = new_embeddings
+
+    def forward(
```
Comment (dlwh) [inline src/levanter/models/roberta.py:778]: in general we use __call__ and not forward in Levanter

File: src/levanter/models/roberta.py [hunk 3]
```diff
@@ -0,0 +1,852 @@
...
+            position_ids=position_ids,
+            inputs_embeds=inputs_embeds
+        )
+
+        sequence_output = outputs[0]
+        prediction_scores = self.lm_head(sequence_output)
+
+        masked_lm_loss = None
+        if labels is not None:
+            masked_lm_loss = hnn.cross_entropy_loss(logits=prediction_scores, Label=self.Vocab, targets=labels)
```
Comment (dlwh) [inline src/levanter/models/roberta.py:809]: my preference is to put this part in a method called compute_loss that calls `__call__` (which computes the logits), and then computes the loss


also you'll want to use the loss mask when that gets merged here

**Review by dlwh — COMMENTED** [review]
didn't check correctness. just some style/organization thoughts

**Comment by dlwh** [comment]
I should add: looking good! I know this ended up being a big lift and I appreciate you all taking it on!

File: src/levanter/models/roberta.py [hunk 4]
```diff
@@ -0,0 +1,852 @@
...
+
+        context_layer = hax.dot(attention_probs, value_layer, axis=self.KeyPos)
+        
+        outputs = hax.rearrange(context_layer, ("... heads position head_size -> ... position (embed_att: heads head_size)"), heads=self.Heads, head_size=self.HeadSize)
+
+        return outputs
+
+class RobertaSelfOutput(eqx.Module, StateDictSerializationMixin):
+    dense: hnn.Linear
+    LayerNorm: hnn.LayerNorm
```
[the remark this answers is not in the document]
Comment (JulienDarve) [reply src/levanter/models/roberta.py:361]: It is named this in the transformers library for most of the functions, so I made it that way here. I originally had it in snake case.

File: src/levanter/models/roberta.py [hunk 5]
```diff
@@ -0,0 +1,852 @@
...
+        return {"self_attn": "self"}
+
+    @named_call
+    def __call__(
+        self,
+        hidden_states: NamedArray,
+        attention_mask: Optional[NamedArray] = None,
+        *,
+        key
+    ) -> Tuple[NamedArray]:
```
[the remark this answers is not in the document]
Comment (JulienDarve) [reply src/levanter/models/roberta.py:403]: Good point

File: src/levanter/models/testing.ipynb
Comment (JulienDarve) [reply src/levanter/models/testing.ipynb:1]: Will do. Do you want me to add it somewhere else in levanter or just keep it to myself?

File: src/levanter/models/roberta.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dlwh) [reply src/levanter/models/roberta.py:361]: totally fine then! probably best

File: src/levanter/models/testing.ipynb
Comment (dlwh) [reply src/levanter/models/testing.ipynb:1]: you can just move it into examples/roberta/ or something. Ideally it was be proper unit tests

File: src/levanter/models/roberta.py [hunk 6]
```diff
@@ -0,0 +1,852 @@
...
+    @named_call
+    def __call__(
+        self,
+        hidden_states: NamedArray,
+        attention_mask: Optional[NamedArray] = None,
+        *,
+        key = None
+    ) -> Tuple[NamedArray]:
+
+        query_layer = self.transpose_for_scores(self.q_proj(hidden_states))
```
[the remark this answers is not in the document]
Comment (dlwh) [reply src/levanter/models/roberta.py:304]: parameters are best declared in the same order as they are in HF (though that can be worked around) but intermediate values like query_layer can be in any order really. Haliax will automatically transpose things as needed.

**Comment by versae** [comment]
With the release of ModernBERT, I was wondering if there are any plans to merge this RoBERTa into levanter and possibly improve it add some nice things from ModernBERT like RoPE.

**Comment by dlwh** [comment]
@versae Yeah I think that's basically the direction they're/we're heading. Basically just use Llama with a different loss function/acausal attention
